Add PKCS#11 + yubihsm-auth integration

myksyr-tdy opened the pull request:
Integrate PKCS#11 library with yubihsm-auth authentication. When prompted by PKCS#11 to enter PIN code you need to enter "#$label#$key_id$pin" where
 - $label is the credential label in yubihsm-auth
 - $key_id is the key ID in the YubiHSM, four hex digits
 - $pin is the credential password in yubihsm-auth

myksyr-tdy commented:
Is build test failure caused by some pipeline issue (curl)?

Tests pass locally as reported by "make test":
100% tests passed, 0 tests failed out of 31

Total Test time (real) = 233.84 sec

The following tests did not run:
	 22 - encrypt_aes (Skipped)
	 25 - aes_encrypt_test (Skipped)

qpernil commented:
AES tests are skipped if the yubihsm doesn't support it (based on the version number)

myksyr-tdy commented:
@qpernil , is there anything I can do to help integrate this PR?

TheButlah commented:
@qpernil YubiHSM customer here. My team is using YubiHSM Auth (ykopen) based asymmetric authentication keys to handle access to the YubiHSM 2. We rely on the PKCS11 interface for integration with vendor specific secure boot code signing tools. 

Without this PR, there is no real way I can use the YubiHSM as the root of trust of a device fleet.

Has there been any progress on integrating this PR?

aveenismail commented:
@TheButlah Apologies for the late response.

We've now moved this PR up the priority list and is now going through the review process.

Once merged, do you need new release binaries or are you building from repository?

myksyr-tdy commented:
Original submitter here—great to see this patch being adopted elsewhere! I’m happy to help adjust the PR as needed. We also use this in a production system, so an SDK release would be preferable to simplify our tooling.

aveenismail commented:
Sorry @myksyr-tdy, I came late into the discussion so I missed who the original submitter is.

Thank you. If possible, can you please re-base it on the current master?

myksyr-tdy committed:
doc: document asymmetric and yubihsm-auth PKCS#11 login formats

myksyr-tdy committed:
pkcs11: add yubihsm-auth support

Integrate PKCS#11 library with yubihsm-auth authentication.

When prompted by PKCS#11 to enter PIN code you need to enter
"#$label#$key_id$pin" where
 - $label is the credential label in yubihsm-auth
 - $key_id is the key ID in the YubiHSM, four hex digits
 - $pin is the credential password in yubihsm-auth

myksyr-tdy commented:
@aveenismail, I've rebased the patch on to of master now. I took care of updating create_yksession to support asymmetric auth which weren't available at the time of original submission, as well as documented both asymmetric and YubiKey-based auth in README.

aveenismail commented:
Thank you @myksyr-tdy, very appreciated. 

I've done now the initial review and your PR actually pointed us to a possible memory leak due to data not being erased after use. I see that this PR is modeled after the implementation of ykopen command in yubihsm-shell so I've done the fix there, just need to run the final tests before pushing it. I'll let you know when I've pushed it and would ask you to apply the fix to this PR. After that, it'll have to go through a separate security review before merging. I don't anticipate much issues there though so hopefully we can merge this into master soon.

TheButlah commented:
@aveenismail thanks for the prioritization! i have been using the nixos packaging for it, although im unsure if its up to date or not. I tried getting things working on the pr a week or so ago and it seem to build with nix, but I'm sure I can fix it on my end once its been merged

aveenismail commented:
Sorry, I must have lost my memory over night. I actually already pushed the fix in this PR https://github.com/Yubico/yubihsm-shell/pull/529/changes. The tests I'm running now are on this PR

Can you please apply the fix similar to the PR above? Notice also that part of the fix is that a session now gets closed if its authentication fails.

aveenismail commented:
Thank you @myksyr-tdy. Sending this now to the security review

aveenismail commented on pkcs11/yubihsm_pkcs11.c:
```diff
@@ -6233,6 +6402,38 @@ CK_DEFINE_FUNCTION(CK_RV, C_LoginUser)
+  CK_UTF8CHAR_PTR pLabel = NULL;
+  CK_ULONG ulLabelLen = 0;
+
+  if (*pUsername == '#') {
```
Please add a check here that ulUsernameLen is at least 2 chars long to avoid having its value wrapped if `ulUsernameLen--` makes it lower than 0

myksyr-tdy replied:
Would ulUsernameLen && *pUsername == '#' work? Then "Can't locate trailing #" will take care of username consisting of single "#".

I'll send and updated patch, but I'm open to changing it to >= 2 too.

aveenismail replied:
That should fix it most of it, yes. Just add a check that `ulUsernameLen` is a positive value going into the parsing. Something like 
`if (pUsername == NULL || ulUsernameLen == 0) {...}` or `CK_UTF8CHAR prefix = ulUsernameLen ? *pUsername : 0`. This should cover it.

The one other finding is that in line 1268 the error YKHSMAUTHR_WRONG_PW but is practically mapped to YHR_GENERIC_ERROR, which can make errors hard to debug. Can you please add a line that sets yrc to something more descriptive, like YHR_DEVICE_AUTHENTICATION_FAILED or YHR_SESSION_AUTHENTICATION_FAILED?

Once these are fixes, I'll start merging the PR

aveenismail commented on pkcs11/yubihsm_pkcs11.c:
Please add a check here that `ulUsernameLen` is at least 1 char long to avoid having its value wrapped if `ulUsernameLen--` makes it lower than 0

aveenismail commented:
@myksyr-tdy The initial security review found a possible Out of Bounds Read if `ulUsernameLen--` puts the value bellow 0 in `C_LoginUser`. Because of the line change in the `C_Login` function, this effects the parsing of the PIN pattern that starts with `@` too. I added review comments for where there need to be a length check.

I'll keep you posted in case there are further findings.

aveenismail commented:
@myksyr-tdy A point of clarification, is there a reason why the PIN pattern prefix would be `@#`? As you stated correctly in the README, when the credential is stored on a YubiKey, the pkcs11 module does not need to be aware whether the authentication workflow would be symmetric or asymmetric. So wouldn't it be enough to have the supported patterns as follows so the relevant prefix would only be either `@` or `#`, not both?

-  {AuthKeyID}{Password} => regular symmetric authentication using a derived password
-  @{AuthKeyID}{Password} => asymmetric authentication 
- #{YkAuthLabel}#{AuthKeyID}{Password} => use credential stored on a YubiKey for authentication.

myksyr-tdy commented:
"@#" is silently accepted as equivalent to "#", but I can rework the code to only accept one of those three formats.

aveenismail commented:
Thank you so much @myksyr-tdy for the PR and all the adjustments, and thank you @TheButlah for the feedback. I'll merge this PR now and will try to get a release out by the end of next week
